doc/design/service-discovery.md

```diff
@@ -0,0 +1,304 @@
+
+Con:
+
+* Overhead, `Listner` will wake up on every `CaproMessage`
+* Bookkeeping at user-side, just delta of service registry is transferred
+* What will RouDi do if he runs out of memory?
+  * Dimensioning according to max values is not optimal (MAX_INTERFACE_CAPRO_FIFO_SIZE)
+  * Presumably lots of memory needed, e.g. during startup phase when lots of apps will do discovery
+  * A safety-certified middleware based on iceoryx would not use the dynamic discovery feature
+
+Remark:
+
+* Need to filter out your own `CaproMessageType::FIND` requests
+* Extend `dispatchCaproMessage()` with condition variable and notification mechanism
+
+##### Alternative C: Custom thread
+
+* Implement a custom thread in `PoshRuntime`, which polls for specific discovery information using request/response
+
+Pro:
+
+* More flexible, e.g. thread can be used for other functionality in future
+* Re-use req/resp building blocks
+
+Con:
+
+* Not needed by all users per default
+* Polling leads to overhead
+
+##### Alternative D: Discovery Class + Listner
+
+Create a new publisher in RouDi which sends a `ServiceRegistryTopic`. This publisher would be used both to signal a
+change in the service registry and to transmit the service discovery registry. The complete old service registry
+(saved locally) would be compared to the new service registry in a new class, extending the public user API.
+
+Pro:
+
+* Simple and consistent user API for both event-based and sychronous requests
+  * Filtering for `findService` could be done inside the new class
+* No need for a change counter as ABA problem solved by POSH mechanisms
+* Not just delta of service registry is transmitted, but complete service registry info
+* Fast data transport
+* Close to DDS mechanism
+* Could replace `InterfacePort`s
+
+Con:
+
+* New publisher in RouDi needed
+* Partially overlapping features with `InterfacePort`s
+  * Could the `InterfacePort`s be replace by alternative D?
+
+### Decision
+
+<!-- @todo Add final decision -->
+
+* Sychronous, one-shot RPC (Polling): Alternative A
+* Event-based notification: Alternative B
+
+### Code example
+
+#### Event-based notification: Alternative B
+
+```cpp
+auto& runtime = PoshRuntime::initRuntime("myApp");
+auto* interfacePortData = runtime.getMiddlewareInterface(capro::Interfaces::INTERNAL);
+
+InterfacePort caproSubscriber(interfacePortData);
+Listener myListner;
+
+void onDiscoveryUpdateCallback(InterfacePort* subscriber)
+{
+    subscriber->tryGetCaProMessage().and_then([](auto& caproMessage){
+        // Has any service that is relevant for me changed?
+        if(caproMessage.m_serviceDescription == myRelevantServiceDescription && someOtherCondition())
+        {
+            // call user-defined callback
+        }
+    });
+}
+
+myListner.attachEvent(caproSubscriber, DATA_RECEIVED, createNotificationCallback(onDiscoveryUpdateCallback));
+```
+
+#### Event-based notification: Alternative D
+
+```cpp
+class DiscoveryInfo
```

Comment:
I favor this alternative but there remains two important open question: How do we ensure that there are enough chunks available?

Availability of chunks:
1. The user provides a custom mempool config where samples with the required size are not available at all. Or when they are available there are not enough of them.
2. The users publisher is using the same samples for user data and we are unable to acquire a chunk to send the service registry update.

Access rights:
The mempool must be inside a shared memory segment which is readable by all. What happens when the user configures a custom segment config where a certain application has no read access and is unable to retrieve the samples.


One solution for the first implementation could be to create a mempool inside the management segment which is used only for this particular publisher which sends the service registry.
Could you please include a design draft for this approach (or another) to solve the above mentioned problems.

Reply:
Yes, I think in this case the memory used for this kind of topic must be separate from all other shared memory, i.e. no other publisher uses this. We need to move in the direction of customizing the memory allocator per publisher anyway I think, and as a default it may behave as it does now for all publishers.

There is possibly no perfect way to ensure sufficient memory is available, even with this approach. I suspect discovery workload will spike on startup (many services offered one after the other leads to frequent changes)  and require the most load. The memory must be chosen to anticipate this workload. The publisher could use the blocking send publish feature to ensure nothing gets lost, at the expense of a delay in the information being provided to all subscribing runtimes.

I do not think transmitting the whole registry every time is efficient and the way to go for larger systems, it must be done incrementally (i.e. we transmit deltas only).

Reply:
Important point! I added a node, that all communication shall be done over the `iceoryx_managment` segment, so that the user can not interfere with it through a wrong configuration. For me this also applies to the introspection publisher, so I think I'll refactor them alongside to publish in the `iceoryx_mgmt` segment.

Reply:
@mossmaurice sounds like a good plan.